File: LeetCode/0232-Implement Queue using Stacks/main.cpp
//
// Created by 薛智钧 on 2020/4/3.
//
#include <stack>
#include <vector>
#include <iostream>
using namespace std;

/*
   Implement the following operations of a queue using stacks.

   push(x) -- Push element x to the back of queue.
   pop() -- Removes the element from in front of queue.
   peek() -- Get the front element.
   empty() -- Return whether the queue is empty.

 * */

// 思路：利用两个stack,新进栈的缓存到_new,当要pop()或者peek()的时候再转移到old
class MyQueue {
private:
    stack<int> _old, _new;
public:
    /** Initialize your data structure here. */
    MyQueue() { }

    /** Push element x to the back of queue. */
    void push(int x) {
        _new.push(x);
    }

    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        int val = peek();
        _old.pop();
        return val;
    }

    /** Get the front element. */
    int peek() {
        shiftStack();
        return _old.top();
    }

    /** Returns whether the queue is empty. */
    bool empty() {
        return _new.empty() && _old.empty();
    }

    void shiftStack(){
        if (!_old.empty()) return;
        while(!_new.empty()){
            _old.push(_new.top());
            _new.pop();
        }
    }
};

int main(){
    MyQueue my_queue;
    int a = 1, b = 3;
    my_queue.push(a);
    my_queue.push(b);
    cout << my_queue.pop() << endl;
    cout << my_queue.peek();
}

File: LeetCode/0232-Implement Queue using Stacks/main.py
"""  
Implement the following operations of a queue using stacks.

push(x) -- Push element x to the back of queue.
pop() -- Removes the element from in front of queue.
peek() -- Get the front element.
empty() -- Return whether the queue is empty.
Example:

MyQueue queue = new MyQueue();

queue.push(1);
queue.push(2);  
queue.peek();  // returns 1
queue.pop();   // returns 1
queue.empty(); // returns false
"""

class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self._new = []
        self._old = []

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self._new.append(x)

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        self.shuffle()
        return self._old.pop()

    def peek(self) -> int:
        """
        Get the front element.
        """
        self.shuffle()
        return self._old[-1]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return not self._new and not self._old

    def shuffle(self):
        if self._old:
            return

        while self._new:
            self._old.append(self._new.pop())
